transport/src/main/java/io/netty/channel/pool/AbstractChannelPoolMap.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright 2015 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.channel.pool;
+
+import io.netty.util.internal.PlatformDependent;
+
+import java.util.Iterator;
+import java.util.Map.Entry;
+import java.util.concurrent.ConcurrentMap;
+
+import static io.netty.util.internal.ObjectUtil.checkNotNull;
+
+/**
+ * A skeletal {@link ChannelPoolMap} implementation. To find the right {@link ChannelPool}
+ * the {@link Object#hashCode()} and {@link Object#hashCode()} is used.
+ */
+public abstract class AbstractChannelPoolMap<K, P extends ChannelPool>
+        implements ChannelPoolMap<K, P>, Iterable<Entry<K, P>> {
+    private final ConcurrentMap<K, P> map = PlatformDependent.newConcurrentHashMap();
+
+    @Override
+    public final P get(K key) {
+        P pool = map.get(checkNotNull(key, "key"));
+        if (pool == null) {
+            pool = newPool(key);
+            P old = map.putIfAbsent(key, pool);
+            if (old != null) {
+                pool = old;
+            }
+        }
+        return pool;
+    }
+
+    /**
+     * Remove the {@link ChannelPool} from this {@link AbstractChannelPoolMap}. Returns {@code true} if removed,
+     * {@code false} otherwise.
+     */
+    public final boolean remove(K key) {
+        return map.remove(checkNotNull(key, "key")) != null;
+    }
+
+    @Override
+    public final Iterator<Entry<K, P>> iterator() {
+        return map.entrySet().iterator();
```

Comment:
Should we allow users to call "remove" on the iterator?  If we need to do cleanup (https://github.com/netty/netty/pull/3607/files#r29299516) this may complicate things.